src/examples/libpmemobj/queue/queue.c

```diff
@@ -0,0 +1,279 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * queue.c -- array based queue example
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <libpmemobj.h>
+
+POBJ_LAYOUT_BEGIN(queue);
+POBJ_LAYOUT_ROOT(queue, struct root);
+POBJ_LAYOUT_TOID(queue, struct entry);
+POBJ_LAYOUT_TOID(queue, struct queue);
+POBJ_LAYOUT_END(queue);
+
+struct entry { /* queue entry that contains arbitrary data */
+	size_t len; /* length of the data buffer */
+	char data[];
+};
+
+struct queue { /* array-based queue container */
+	size_t front; /* position of the first entry */
+	size_t rear; /* position of the last entry */
+
+	size_t capacity; /* size of the entries array, must be power of two */
+	TOID(struct entry) entries[];
+};
+
+struct root {
+	TOID(struct queue) queue;
+};
+
+/*
+ * queue_constructor -- constructor of the queue container
+ */
+static int
+queue_constructor(PMEMobjpool *pop, void *ptr, void *arg)
+{
+	struct queue *q = ptr;
+	size_t *capacity = arg;
+	q->front = 0;
+	q->rear = 0;
+	q->capacity = *capacity;
+
+	/* atomic API requires that objects are persisted in the constructor */
+	pmemobj_persist(pop, q, sizeof(*q));
+
+	return 0;
+}
+
+/*
+ * queue_new -- allocates a new queue container using the atomic API
+ */
+static int
+queue_new(PMEMobjpool *pop, TOID(struct queue) *q, size_t nentries)
+{
+	return POBJ_ALLOC(pop, q, struct queue,
+		sizeof(struct queue) + sizeof(TOID(struct entry)) * nentries,
+		queue_constructor, &nentries);
+}
+
+/*
+ * queue_nentries -- returns the number of entries
+ */
+static size_t
+queue_nentries(struct queue *queue)
+{
+	return queue->rear - queue->front;
+}
+
+/*
+ * queue_push -- allocates and inserts a new entry into the queue
+ */
+static int
+queue_push(PMEMobjpool *pop, struct queue *queue, char *data, size_t len)
+{
+	if (queue->capacity - queue_nentries(queue) == 0)
+		return -1; /* at capacity */
+
+	/* rear is never decreased, need to calculate the real position */
+	size_t pos = queue->rear & (queue->capacity - 1);
+
+	int ret = 0;
+
+	TX_BEGIN(pop) {
+		/* let's first reserve the space at the end of the queue */
+		TX_ADD_FIELD_DIRECT(queue, rear);
```

Comment:
My personal opinion is that when \_FIELD\_ macros are misdesigned. The next line is "queue->rear += 1", so I think "TX_ADD_DIRECT(&queue->rear)" is easier to understand.

Reply:
done